### PR TITLE
Classes, Interface and API contracts for telemetry

### PR DESCRIPTION
**Pull request by addev-ashish**
A first level PR to make sure that code across various languages follows similar API contracts for telemetry

### CODE REVIEW

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    @Override
+    public List<Pair<String, String>> getEvents() {
+        return mEventList;
+    }
+
+    @Override
+    public int getDefaultEventCount() {
+        return mDefaultEventCount;
+    }
+}
```
Comment (abhishek58g) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: Please consider removing mDefaultCount....It doesnt feature in the latest design

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        mEventList.add(Pair.create("Event", eventName));
+    }
+}
+
+class GrantEvent extends DefaultEvent{
+    GrantEvent(final String eventName) {
+        mEventList.add(Pair.create("Event", eventName));
+    }
+}
+
```
Comment (abhishek58g) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: Can you create separate class files ?...Isnt it a more preferred code design ?

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.List;
+
+interface EventsInterface {
+    void setEvent(String name, String value);
+    List<Pair<String,String>> getEvents();
+    int getDefaultEventCount();
+}
+
+class DefaultEvent implements EventsInterface {
+    List<Pair<String,String>> mEventList = null;
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: >List<Pair<String,String>> mEventList = null; [](start = 4, length = 44)

I think you should remove the =null here, and make it final, initialize it in the constructor.

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    @Override
+    public void setEvent(String name, String value) {
+        if (!TextUtils.isEmpty(name)) {
+            mEventList.add(Pair.create(name, value));
+        }
+    }
+
+    @Override
+    public List<Pair<String, String>> getEvents() {
+        return mEventList;
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: we should return immutable list.  You only expect caller to do read operations, right?

File: adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.aad.adal;
+
+import java.util.List;
+import java.util.Map;
+
+class DefaultDispatcher {
+    Map <String, List<EventsInterface>> mObjectsToBeDispatched;
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java]: >Map <String, List<EventsInterface>> mObjectsToBeDispatched; [](start = 4, length = 59)

no space after Map. 
Also, why can't we make this a final, and initialize it in  the constructor?

File: adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    Dispatcher mDispatcher;
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java]: >Dispatcher mDispatcher; [](start = 4, length = 23)

final

File: adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    DefaultDispatcher() {
+        mDispatcher = null;
+    }
+    DefaultDispatcher(Dispatcher dispatcher) {
+        mDispatcher = dispatcher;
+    }
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java]: nit: extra line

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    @Override
+    public int getDefaultEventCount() {
+        return mDefaultEventCount;
+    }
+}
+
+class HttpEvent extends DefaultEvent{
+    HttpEvent(final String eventName) {
+        mEventList.add(Pair.create("Event", eventName));
+    }
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: why this cannot go to the default event? it's the same across all the three events. 
also, when these each individual concrete event is initialized, shouldn't we already know the event name? why wer're passing it in the constructor? we should just do a new HttpEvent().

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+package com.microsoft.aad.adal;
+
+import android.text.TextUtils;
+import android.util.Pair;
+
+import java.util.ArrayList;
+import java.util.List;
+
+interface EventsInterface {
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: Java interface usually starts with I. We should name it as IEvents.

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Override
+    public int getDefaultEventCount() {
+        return mDefaultEventCount;
+    }
+}
+
+class HttpEvent extends DefaultEvent{
+    HttpEvent(final String eventName) {
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: >final String eventName [](start = 14, length = 22)

:) this is not consistent across the files, if we're going to use final, then please make sure you have it in all the code you add.

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+interface EventsInterface {
+    void setEvent(String name, String value);
+    List<Pair<String,String>> getEvents();
+    int getDefaultEventCount();
+}
+
+class DefaultEvent implements EventsInterface {
+    List<Pair<String,String>> mEventList = null;
+    int mDefaultEventCount = 0;
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: >mDefaultEventCount [](start = 8, length = 18)

can you remind me what this mDefaultEventCount is used for and why it is 0 in the default event.

File: adal/src/main/java/com/microsoft/aad/adal/Telemetry.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+    void startEvent(final String requestId, final String eventName) {
+
+    }
+
+    void stopEvent(final String requestId, final EventsInterface events) {
+
+    }
+
+}
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/Telemetry.java]: nit: extra line

File: adal/src/main/java/com/microsoft/aad/adal/Dispatcher.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.aad.adal;
+
+import android.util.Pair;
+
+import java.util.List;
+
+public interface Dispatcher {
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/Dispatcher.java]: >Dispatcher [](start = 17, length = 10)

I would like us to follow the java standardard for naming interface.

Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/Dispatcher.java]: agree

---
In reply to: [71636946](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/686#discussion_r71636946) [](ancestors = 71636946)

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java (hunk 11 above)
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: we dont have the number yet :)

---
In reply to: [71635432](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/686#discussion_r71635432) [](ancestors = 71635432)

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java (hunk 2 above)
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: will do when i send out the full code. not now. 

---
In reply to: [71445671](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/686#discussion_r71445671) [](ancestors = 71445671)

Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: Why not doing it now? It gives a more clear class layout. 

---
In reply to: [71761318](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/686#discussion_r71761318) [](ancestors = 71761318,71445671)

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    }
+}
+
+class UIEvent extends DefaultEvent{
+    UIEvent(final String eventName) {
+        mEventList.add(Pair.create("Event", eventName));
+    }
+}
+
+class GrantEvent extends DefaultEvent{
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: >GrantEvent [](start = 6, length = 10)

Shouldn't we rename as we discussed yesterday?

Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: doing it for now, but will wait for other peoples feedback.

---
In reply to: [71761822](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/686#discussion_r71761822) [](ancestors = 71761822)

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java (hunk 10 above)
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: agree

---
In reply to: [71633784](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/686#discussion_r71633784) [](ancestors = 71633784)

File: adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java (hunk 6 above)
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java]: yes

---
In reply to: [71627297](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/686#discussion_r71627297) [](ancestors = 71627297)

File: adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java (hunk 2 above)
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: i will, just not in this PR. will change the structure and send out a new PR.

---
In reply to: [71761723](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/686#discussion_r71761723) [](ancestors = 71761723,71761318,71445671)

Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultEvent.java]: fair~

---
In reply to: [71800587](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/686#discussion_r71800587) [](ancestors = 71800587,71761723,71761318,71445671)

File: adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+package com.microsoft.aad.adal;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+class DefaultDispatcher {
+    final Map<String, List<IEvents>> mObjectsToBeDispatched = new HashMap<String, List<IEvents>>();
+    IDispatcher mDispatcher;
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultDispatcher.java]: for this two, can we do:
final Map mObjectsToBeDispatched;
final IDispatcher mDispatcher;

all the initialization go into the constructor.

**Comment by weijjia** [comment]
:shipit: after fixing the new comment.

**Pull request closed without merging**